vpr/src/place/weighted_median_move_generator.cpp

```diff
@@ -0,0 +1,178 @@
+#include "weighted_median_move_generator.h"
+#include "globals.h"
+#include <algorithm>
+#include "math.h"
+
+static void get_bb_cost_for_net_excluding_block(ClusterNetId net_id, t_bb_cost* coords, ClusterBlockId block_id, ClusterPinId moving_pin_id, const PlacerCriticalities* criticalities, bool& skip_net);
+
+e_create_move WeightedMedianMoveGenerator::propose_move(t_pl_blocks_to_be_moved& blocks_affected, e_move_type& /*move_type*/, float rlim, const t_placer_opts& placer_opts, const PlacerCriticalities* criticalities) {
+    auto& place_ctx = g_vpr_ctx.placement();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+
+    auto& place_move_ctx = g_placer_ctx.mutable_move();
+
+    /* Pick a random block to be swapped with another random block.   */
+    ClusterBlockId b_from = pick_from_block();
+
+    if (!b_from) {
+        return e_create_move::ABORT; //No movable block found
+    }
+
+    t_pl_loc from = place_ctx.block_locs[b_from].loc;
+    auto cluster_from_type = cluster_ctx.clb_nlist.block_type(b_from);
+    auto grid_from_type = g_vpr_ctx.device().grid[from.x][from.y].type;
+    VTR_ASSERT(is_tile_compatible(grid_from_type, cluster_from_type));
+
+    /* Calculate the Edge weighted median region */
+    t_pl_loc to;
+
+    t_bb_cost coords;
+    t_bb limit_coords;
+
+    //clear the vectors that saves X & Y coords
+    //reused to save allocation time
+    place_move_ctx.X_coord.clear();
+    place_move_ctx.Y_coord.clear();
+
+    //true if the net is a feedback from the block to itself
+    bool skip_net;
+
+    //iterate over block pins
+    for (ClusterPinId pin_id : cluster_ctx.clb_nlist.block_pins(b_from)) {
+        ClusterNetId net_id = cluster_ctx.clb_nlist.pin_net(pin_id);
+        if (cluster_ctx.clb_nlist.net_is_ignored(net_id))
+            continue;
+        if (int(cluster_ctx.clb_nlist.net_pins(net_id).size()) > placer_opts.place_high_fanout_net)
+            continue;
+        //all net pins are weighted with their own crititcalities
+        get_bb_cost_for_net_excluding_block(net_id, &coords, b_from, pin_id, criticalities, skip_net);
+        if (skip_net)
+            continue;
+
+        place_move_ctx.X_coord.insert(place_move_ctx.X_coord.end(), ceil(coords.xmin.criticality * 10), coords.xmin.edge);
```

Comment:
Should define a constant instead of using the magic number 10 (e.g. CRIT_MULT_FOR_MEDIAN). Explain why you have to multiply by this number in a comment.

Reply:
Done by defining a new constant and added a comment explaining the algorithm behind using it.